### PR TITLE
Fix #132

### LINKED ISSUE #132: Panic: attempt to subtract with overflow
Decoding the following input file crashes in debug mode (but not in release mode, because overflows silently wrap in release mode).

[id:000000,sig:06,src:000000,op:flip2,pos:1027.tar.gz](https://github.com/image-rs/jpeg-decoder/files/4192738/id.000000.sig.06.src.000000.op.flip2.pos.1027.tar.gz)

```
$ RUST_BACKTRACE=1 cargo run --bin reproduce_decode out/crashes/id\:000000\,sig\:06\,src\:000000\,op\:flip2\,pos\:1027
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/reproduce_decode 'out/crashes/id:000000,sig:06,src:000000,op:flip2,pos:1027'`
thread 'main' panicked at 'attempt to subtract with overflow', /home/danilo/Projects/jpeg-decoder/src/decoder.rs:760:17
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:61
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1028
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1412
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:65
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:50
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:188
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:205
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:464
  11: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:373
  12: rust_begin_unwind
             at src/libstd/panicking.rs:302
  13: core::panicking::panic_fmt
             at src/libcore/panicking.rs:139
  14: core::panicking::panic
             at src/libcore/panicking.rs:70
  15: jpeg_decoder::decoder::refine_non_zeroes
             at /home/danilo/Projects/jpeg-decoder/src/decoder.rs:760
  16: jpeg_decoder::decoder::decode_block_successive_approximation
             at /home/danilo/Projects/jpeg-decoder/src/decoder.rs:721
  17: jpeg_decoder::decoder::Decoder<R>::decode_scan
             at /home/danilo/Projects/jpeg-decoder/src/decoder.rs:484
  18: jpeg_decoder::decoder::Decoder<R>::decode_internal
             at /home/danilo/Projects/jpeg-decoder/src/decoder.rs:235
  19: jpeg_decoder::decoder::Decoder<R>::decode
             at /home/danilo/Projects/jpeg-decoder/src/decoder.rs:138
  20: reproduce_decode::decode
             at src/reproduce_decode.rs:6
  21: reproduce_decode::main
             at src/reproduce_decode.rs:17
  22: std::rt::lang_start::{{closure}}
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/rt.rs:61
  23: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:48
  24: std::panicking::try::do_call
             at src/libstd/panicking.rs:287
  25: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  26: std::panicking::try
             at src/libstd/panicking.rs:265
  27: std::panic::catch_unwind
             at src/libstd/panic.rs:396
  28: std::rt::lang_start_internal
             at src/libstd/rt.rs:47
  29: std::rt::lang_start
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/rt.rs:61
  30: main
  31: __libc_start_main
  32: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Found using afl (see #131).

Potentially related to #63.

### PR DESCRIPTION
**Pull request by lovasoa**
- Fixes #132
 - improve readability
 - improve performance

 Criterion report on my computer:

```
 Benchmarking decode a 512x512 progressive JPEG: Collecting 100 samples in estimated 27.108 s (5050 i                                                                                                    decode a 512x512 progressive JPEG
                         time:   [5.3440 ms 5.3549 ms 5.3669 ms]
                         change: [-24.307% -23.085% -21.920%] (p = 0.00 < 0.05)
                         Performance has improved.
 Found 3 outliers among 100 measurements (3.00%)
   2 (2.00%) high mild
   1 (1.00%) high severe
```

### CODE REVIEW

**Comment by fintelia** [comment]
Just to confirm, are you sure that the logical OR operation what this code is supposed to be doing? I posted that as a guess in the issue thread, but I don't actually know how jpeg decoding works so I was just speculating based on the shape of the surrounding code.

**Comment by lovasoa** [comment]
No, I did not check... The specification for the part in the code in in G 1.2.3 of the JPEG specification, if you want to have a look yourself :  https://books.google.fr/books?id=AepB_PZ_WMkC&pg=PA484&lpg=PA484&dq=Jpeg+Section+G.1.2.3&source=bl&ots=USKKer5RuP&sig=ACfU3U27xg2SO6Xh3LzRCXmCkqZxwFmspg&hl=en&sa=X&ved=2ahUKEwjv4bb11s7nAhUDXqwKHXnjClQQ6AEwCnoECAsQAQ#v=onepage&q&f=false

**Comment by lovasoa** [comment]
I also gave maintainers edit rights on this pull request, so if you want to fix something, you can commit directly on lovasoa:master if you want

File: src/decoder.rs [hunk 1]
```diff
@@ -739,26 +739,19 @@ fn refine_non_zeroes<R: Read>(reader: &mut R,
                               bit: i16) -> Result<u8> {
     debug_assert_eq!(coefficients.len(), 64);
 
-    let last = range.end - 1;
+    let mut last= range.start;
```
Comment (dbrgn) [inline src/decoder.rs]: Nitpick: Missing whitespace.

```suggestion
    let mut last = range.start;
```

**Comment by lovasoa** [comment]
I added the whitespace. But what I think we need is someone reviewing that against the JPEG specification.

**Pull request closed without merging**